buildSrc/src/main/java/org/elasticsearch/gradle/PublishPlugin.java

```diff
@@ -0,0 +1,202 @@
+                project.getConfigurations().getByName(ShadowBasePlugin.getCONFIGURATION_NAME()).getAllDependencies().all(dependency -> {
+                    if (dependency instanceof ProjectDependency) {
+                        Node dependencyNode = dependenciesNode.appendNode("dependency");
+                        dependencyNode.appendNode("groupId", dependency.getGroup());
+                        ProjectDependency projectDependency = (ProjectDependency) dependency;
+                        String artifactId = getArchivesBaseName(projectDependency.getDependencyProject());
+                        dependencyNode.appendNode("artifactId", artifactId);
+                        dependencyNode.appendNode("version", dependency.getVersion());
+                        dependencyNode.appendNode("scope", "compile");
+                    }
+                });
+            });
+        });
+
+        // Add git origin info to generated POM files
+        publishing.getPublications().withType(MavenPublication.class, publication -> {
+            publication.getPom().withXml(PublishPlugin::addScmInfo);
+
+            // have to defer this until archivesBaseName is set
+            project.afterEvaluate(p -> publication.setArtifactId(getArchivesBaseName(project)));
+
+            generatePomTask.configure(
+                t -> t.dependsOn(String.format("generatePomFileFor%sPublication", Util.capitalize(publication.getName())))
+            );
+        });
+
+    }
+
+    private static void addScmInfo(XmlProvider xml) {
+        Node root = xml.asNode();
+        root.appendNode("url", Util.urlFromOrigin(BuildParams.getGitOrigin()));
+        Node scmNode = root.appendNode("scm");
+        scmNode.appendNode("url", BuildParams.getGitOrigin());
+    }
+
+    private static void configureJavadoc(Project project) {
+        // remove compiled classes from the Javadoc classpath: http://mail.openjdk.java.net/pipermail/javadoc-dev/2018-January/000400.html
+        final List<File> classes = new ArrayList<>();
+        project.getTasks().withType(JavaCompile.class).configureEach(javaCompile -> classes.add(javaCompile.getDestinationDir()));
+        project.getTasks().withType(Javadoc.class).configureEach(javadoc -> {
+            // only explicitly set javadoc executable if compiler JDK is different from Gradle
+            // this ensures better cacheability as setting ths input to an absolute path breaks portability
+            Path compilerJvm = BuildParams.getCompilerJavaHome().toPath();
+            Path gradleJvm = Jvm.current().getJavaHome().toPath();
+            try {
+                if (Files.isSameFile(compilerJvm, gradleJvm) == false) {
+                    javadoc.setExecutable(compilerJvm.resolve("bin/javadoc").toString());
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+            javadoc.setClasspath(javadoc.getClasspath().filter(f -> classes.contains(f) == false));
```

Comment:
We can simplify this a lot. If all we want is the compile class minus the source set output we can just use the main source set's `getCompileClasspath()`.

Reply:
I think this is a bit different. See http://mail.openjdk.java.net/pipermail/javadoc-dev/2018-January/000400.html

Reply:
Oh I see what you mean now. Done.